### DISCUSSION TITLE
Can @TestTransaction be made to work with Jupiter @BeforeEach?

### DISCUSSION THREAD

**Discussion by franck102** [quarkusio/quarkus#40119] [Q&A]
Using the latest Quarkus and Junit Jupiter I have a large set of service tests that creates test fixtures in the DB. I am using @TestTransaction to make sure that the fixtures don't persist in the DB once the tests complete.

This pattern works as expected:

```
@QuarkusTest
public class BookingTests extends BookingTestBase
{
    @Test
    @TestTransaction
    public void bookSession() throws IOException
    {
        createFixtures();
        ... test some stuff;
    }
}
```

however I cannot figure out how I could pull the createFixtures call in an @BeforeEach method:

```
@QuarkusTest
public class BookingTests extends BookingTestBase
{

    @BeforeEach
    public void beforeEach() {
         createFixtures();
    }
 
    @Test
    @TestTransaction
    public void bookSession() throws IOException
    {
        ... test some stuff;
    }
}
```
With the code above the fixtures persist after the tests.
I have tried moving the @TestTransaction to the class level, adding it to the @BeforeEach method, nothing helps... any suggestion?

Thanks!
Franck

[Edit]: Related question: is there any simple way to check whether I am currently in the context of an @TestTransaction? What I really need to do is make sure I don't allow createFixtures() to proceed unless I know it will be rolled back...

**Comment by kozaka-tv** [comment]
I am facing with the same Problem when I am using an Abstract class. That the @BeforeEach does have probably a different context, so the Transaction is then gone. And in the test method, I do not have the database changes made in createFixtures().

**Comment by MaklTa** [comment]
any chance to solve this with the refactored testing feature introduced in 3.22 @holly-cummins?
